domain/port/state/state.go

```diff
@@ -0,0 +1,271 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/set"
+	"github.com/juju/collections/transform"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/domain"
+	internaldatabase "github.com/juju/juju/internal/database"
+	"github.com/juju/juju/internal/uuid"
+)
+
+// State represents the persistence layer for opened ports.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new state reference.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// GetOpenedPorts returns the opened ports for a given unit,
+// grouped by endpoint.
+func (st *State) GetOpenedPorts(ctx context.Context, unitUUID string) (network.GroupedPortRanges, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	m := sqlair.M{
+		"unitUUID": unitUUID,
+	}
+
+	getOpenedPorts, err := st.Prepare(`
```

Comment:
Just call this query. I think this name is confusing, especially when you have a method with the same name.

Reply:
Yeah fair. It was only late in development that this method was reduced to a single query, so this is ok